lib/cli/src/Cardano/CLI.hs

```diff
@@ -367,114 +377,139 @@ runCli = join . customExecParser preferences
-    decode txt = do
-        bytes <- fromHex $ B8.pack $ T.unpack . T.strip $ txt
-        case BS.length bytes of
-            96 -> fmap AXPrv $ left showErr $ xPrvFromStrippedPubXPrvCheckRoundtrip bytes
-            64 -> fmap AXPub . CC.xpub $ bytes
-            n -> Left . mconcat $
-                [ "Expected key to be 96 bytes in the case of a private key"
-                , " and, 64 bytes for public keys. This key is "
-                , show n
-                , " bytes."
-                ]
-      where
-        showErr (ErrInputLengthMismatch expected actual) = mconcat
-            [ "Expected extended private key to be "
-            , show expected
-            , " bytes but got "
-            , show actual
+keyFromBytes :: ByteString -> Either String XPrvOrXPub
+keyFromBytes bytes = do
+    case BS.length bytes of
+        96 -> fmap AXPrv $ left showErr $ xPrvFromStrippedPubXPrvCheckRoundtrip bytes
+        64 -> fmap AXPub . CC.xpub $ bytes
+        n -> Left . mconcat $
+            [ "Expected key to be 96 bytes in the case of a private key"
+            , " and, 64 bytes for public keys. This key is "
+            , show n
             , " bytes."
             ]
-        showErr (ErrCannotRoundtripToSameBytes) = mconcat
-            [ "That extended private key looks weird. "
-            , "Is it encrypted? Or is it an old Byron key?"
-            ]
-        fromHex = left (const "Invalid hex.")
-            . convertFromBase Base16
+  where
+    showErr (ErrInputLengthMismatch expected actual) = mconcat
+        [ "Expected extended private key to be "
+        , show expected
+        , " bytes but got "
+        , show actual
+        , " bytes."
+        ]
+    showErr (ErrCannotRoundtripToSameBytes) = mconcat
+        [ "That extended private key looks weird. "
+        , "Is it encrypted? Or is it an old Byron key?"
+        ]
+
+xprvHrp :: Bech32.HumanReadablePart
+xprvHrp = [Bech32.humanReadablePart|xprv|]
+
+xpubHrp :: Bech32.HumanReadablePart
+xpubHrp = [Bech32.humanReadablePart|xpub|]
+
+decodeAnyKey :: Text -> Either String (XPrvOrXPub, KeyEncoding)
+decodeAnyKey rawKey = do
+    enc <- detectEncoding rawKey
+    k <- decodeKey enc rawKey
+    return (k, enc)
+
+decodeKey :: KeyEncoding -> Text -> Either String XPrvOrXPub
+decodeKey = \case
+    Bech32 -> decodeBech32
+    Hex -> decodeHex >=> keyFromBytes
+  where
+    decodeBech32 :: Text -> Either String XPrvOrXPub
+    decodeBech32 t = do
+        (hrp, dp) <- left show $ Bech32.decodeLenient t
+        bytes <- maybe (Left dpErr) Right $ Bech32.dataPartToBytes dp
+        case hrp of
+            h | h == xpubHrp -> keyFromBytes bytes
+            h | h == xprvHrp -> keyFromBytes bytes
+            _ -> Left "unrecognized Bech32 Human Readable Part"
+      where
+        dpErr = "Internal error: Unable to extract bytes from bech32 data part"
+
+decodeHex :: Text -> Either String ByteString
+decodeHex txt = fromHex $ B8.pack $ T.unpack . T.strip $ txt
```

Comment:
This works because hex-encoded characters are ASCII characters, but the "good way" of doing this would be by using `encodeUtf8` from `Data.Text.Encoding`.